// selectn/selectn.js
!function(e){if("object"==typeof exports&&"undefined"!=typeof module)module.exports=e();else if("function"==typeof define&&define.amd)define([],e);else{var f;"undefined"!=typeof window?f=window:"undefined"!=typeof global?f=global:"undefined"!=typeof self&&(f=self),f.selectn=e()}}(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict'

/**
 * imports.
 */

var dotted = require('brackets2dots')
var splits = require('dotsplit.js')
var curry2 = require('curry2')
var string = Object.prototype.toString

/**
 * exports.
 */

module.exports = curry2(selectn)

/**
 * Curried property accessor function that resolves deeply-nested object properties via dot/bracket-notation
 * string path while mitigating `TypeErrors` via friendly and composable API.
 *
 * @param {String|Array} path
 * Dot/bracket-notation string path or array.
 *
 * @param {Object} object
 * Object to access.
 *
 * @return {Function|*|undefined}
 * (1) returns `selectn/1` when partially applied.
 * (2) returns value at path if path exists.
 * (3) returns undefined if path does not exist.
 */

function selectn (path, object) {
  var idx = -1
  var seg = string.call(path) === '[object Array]' ? path : splits(dotted(path))
  var end = seg.length
  var ref = end ? object : void 0

  while (++idx < end) {
    if (Object(ref) !== ref) return void 0
    ref = ref[seg[idx]]
  }

  return typeof ref === 'function' ? ref() : ref
}

},{"brackets2dots":6,"curry2":7,"dotsplit.js":8}],2:[function(require,module,exports){
'use strict'

/*!
 * imports.
 */

var curry2 = require('curry2')
var selectn = require('selectn')

/*!
 * imports (local).
 */

var expressions = require('./lib/expressions')

/*!
 * exports.
 */

module.exports = curry2(filter)

/**
 * Curried function deriving a new array containing items from given array for which predicate returns true.
 * Supports unary function, RegExp, dot/bracket-notation string path, and compound query object as predicate.
 *
 * @param {Function|String|Object} predicate
 * Unary function, RegExp, dot/bracket-notation string path, and compound query object.
 *
 * @param {Array} list
 * Array to evaluate.
 *
 * @return {Array}
 * New array containing items from given array for which predicate returns true.
 */

function filter (predicate, list) {
  var end = list.length
  var idx = -1
  var out = []

  while (++idx < end) {
    if (matchall(expressions(predicate, list[idx]))) out.push(list[idx])
  }

  return out
}

/**
 * Whether all given expressions evaluate to true.
 *
 * @param {Array} expressions
 * Expressions to evaluate.
 *
 * @return {Boolean}
 * Whether all given expressions evaluate to true.
 */

function matchall (expressions) {
  var end = expressions.length
  var idx = -1
  var out = false

  while (++idx < end) {
    var expression = expressions[idx]

    if (({}).toString.call(expression.predicate) === '[object Function]') {
      out = !!expression.predicate(expression.element)
    } else if (({}).toString.call(expression.predicate) === '[object RegExp]') {
      out = !!expression.predicate.exec(expression.element)
    } else if (expression.compare) {
      out = expression.predicate === expression.element
    } else {
      out = selectn(expression.predicate, expression.element)
    }

    if (out === false) {
      return out
    }
  }

  return out
}

},{"./lib/expressions":3,"curry2":7,"selectn":10}],3:[function(require,module,exports){
'use strict'

/*!
 * imports.
 */

var selectn = require('selectn')

/*!
 * exports.
 */

module.exports = expressions

/**
 * Creates and returns an array of expression objects.
 *
 * Example:
 *
 *  {
 *    predicate: 'received',
 *    element: 'received',
 *    compare: true
 *  }
 *
 *  {
 *    predicate: isBoolean,
 *    element: true
 *  }
 *
 *  {
 *    predicate: 'message.read',
 *    element: { message: { read: true } }
 *  }
 *
 * @param {Function|String|Object} predicate
 * Unary function, RegExp, dot/bracket-notation string path, and compound query object.
 *
 * @param {Array} list
 * Array to iterate over.
 *
 * @return {Array}
 * New array containing items from given array for which predicate returns true.
 */

function expressions (predicate, element) {
  var expressions = []

  if (isFunction(predicate) || isRegExp(predicate) || isString(predicate)) {
    expressions.push({predicate: predicate, element: element})
  } else {
    for (var key in predicate) {
      expressions.push({predicate: predicate[key], element: selectn(key, element), compare: true})
    }
  }

  return expressions
}

/**
 * Whether predicate is a RegExp instance.
 *
 * @param {*} predicate
 * Predicate value to test.
 *
 * @return {Boolean}
 * Whether predicate is a RegExp instance.
 */

function isRegExp (predicate) {
  return ({}).toString.call(predicate) === '[object RegExp]'
}

/**
 * Whether predicate is a function.
 *
 * @param {*} predicate
 * Predicate value to test.
 *
 * @return {Boolean}
 * Whether predicate is a function.
 */

function isFunction (predicate) {
  return ({}).toString.call(predicate) === '[object Function]'
}

/**
 * Whether predicate is a string.
 *
 * @param {*} predicate
 * Predicate value to test.
 *
 * @return {Boolean}
 * Whether predicate is a string.
 */

function isString (predicate) {
  return ({}).toString.call(predicate) === '[object String]'
}

},{"selectn":10}],4:[function(require,module,exports){
'use strict'

/*!
 * imports.
 */

var curry2 = require('curry2')
var selectn = require('selectn')

/*!
 * exports.
 */

module.exports = curry2(map)

/**
 * Curried function deriving new array values by applying provided function to each item/index of provided array.
 * Optionally, a dot-notation formatted string may be provided for item property access.
 *
 * @param {Function|String} fn
 * Function to apply to each item.
 *
 * @param {Array} list
 * Array to iterate over.
 *
 * @return {Array}
 * Array resulting from applying provided function `fn` to each item of `list`.
 */

function map (fn, list) {
  var end = list.length
  var idx = -1
  var out = []

  while (++idx < end) {
    out.push((typeof fn === 'string') ? selectn(fn, list[idx]) : fn(list[idx]))
  }

  return out
}

},{"curry2":5,"selectn":10}],5:[function(require,module,exports){
'use strict'

/*!
 * exports.
 */

module.exports = curry2

/**
 * Curry a binary function.
 *
 * @param {Function} fn
 * Binary function to curry.
 *
 * @param {Object} [self]
 * Function `this` context.
 *
 * @return {Function|*}
 * If partially applied, return unary function, otherwise, return result of full application.
 */

function curry2 (fn, self) {
  var out = function () {
    return arguments.length > 1
    ? fn.call(self, arguments[0], arguments[1])
    : fn.bind(self, arguments[0])
  }

  out.uncurry = function uncurry () {
    return fn
  }

  return out
}

},{}],6:[function(require,module,exports){
'use strict';

/*!
 * exports.
 */

module.exports = brackets2dots;

/*!
 * regexp patterns.
 */

var REPLACE_BRACKETS = /\[([^\[\]]+)\]/g;
var LFT_RT_TRIM_DOTS = /^[.]*|[.]*$/g;

/**
 * Convert string with bracket notation to dot property notation.
 *
 * ### Examples:
 *
 *      brackets2dots('group[0].section.a.seat[3]')
 *      //=> 'group.0.section.a.seat.3'
 *
 *      brackets2dots('[0].section.a.seat[3]')
 *      //=> '0.section.a.seat.3'
 *
 *      brackets2dots('people[*].age')
 *      //=> 'people.*.age'
 *
 * @param  {String} string
 * original string
 *
 * @return {String}
 * dot/bracket-notation string
 */

function brackets2dots(string) {
  return ({}).toString.call(string) == '[object String]'
       ? string.replace(REPLACE_BRACKETS, '.$1').replace(LFT_RT_TRIM_DOTS, '')
       : ''
}

},{}],7:[function(require,module,exports){
'use strict'

/*!
 * imports.
 */

if (!Function.prototype.bind) Function.bind = require('fast-bind')

/*!
 * exports.
 */

module.exports = curry2

/**
 * Curry a binary function.
 *
 * @param {Function} fn
 * Binary function to curry.
 *
 * @param {Object} [self]
 * Function `this` context.
 *
 * @return {Function|*}
 * If partially applied, return unary function, otherwise, return result of full application.
 */

function curry2 (fn, self) {
  var out = function () {
    if (arguments.length === 0) return out

    return arguments.length > 1
      ? fn.apply(self, arguments)
      : fn.bind(self, arguments[0])
  }

  out.uncurry = function uncurry () {
    return fn
  }

  return out
}

},{"fast-bind":9}],8:[function(require,module,exports){
'use strict'

/*!
 * imports.
 */

var dotted = require('arraymap')(todots)
var compact = require('array.filter')(String)
var toString = Object.prototype.toString

/*!
 * exports.
 */

module.exports = dotsplit

/**
 * Transform dot-delimited strings to array of strings.
 *
 * @param  {String} string
 * Dot-delimited string.
 *
 * @return {Array}
 * Array of strings.
 */

function dotsplit (string) {
  return dotted(normalize(string))
}

/**
 * Normalize string by:
 *
 * (1) Dropping falsey values (empty, null, etc.)
 * (2) Replacing escapes with a placeholder.
 * (3) Splitting string on `.` delimiter.
 * (4) Dropping empty values from resulting array.
 *
 * @param  {String} string
 * Dot-delimited string.
 *
 * @return {Array}
 * Array of strings.
 */

function normalize (string) {
  return compact(
    (toString.call(string) === '[object String]' ? string : '')
    .replace(/\\\./g, '\uffff')
    .split('.')
  )
}

/**
 * Change placeholder to dots.
 *
 * @param  {String} string
 * Dot-delimited string with placeholders.
 *
 * @return {String}
 * Dot-delimited string without placeholders.
 */

function todots (string) {
  return string.replace(/\uffff/g, '.')
}

},{"array.filter":2,"arraymap":4}],9:[function(require,module,exports){
'use strict';
module.exports = function(boundThis) {
  var f = this
    , ret

  if (arguments.length < 2)
    ret = function() {
      if (this instanceof ret) {
        var ret_ = f.apply(this, arguments)
        return Object(ret_) === ret_
          ? ret_
          : this
      }
      else
        return f.apply(boundThis, arguments)
    }
  else {
    var boundArgs = new Array(arguments.length - 1)
    for (var i = 1; i < arguments.length; i++)
      boundArgs[i - 1] = arguments[i]

    ret = function() {
      var boundLen = boundArgs.length
        , args = new Array(boundLen + arguments.length)
        , i
      for (i = 0; i < boundLen; i++)
        args[i] = boundArgs[i]
      for (i = 0; i < arguments.length; i++)
        args[boundLen + i] = arguments[i]

      if (this instanceof ret) {
        var ret_ = f.apply(this, args)
        return Object(ret_) === ret_
          ? ret_
          : this
      }
      else
        return f.apply(boundThis, args)
    }
  }

  ret.prototype = f.prototype
  return ret
}

},{}],10:[function(require,module,exports){
/*!
 * exports.
 */

module.exports = selectn;

/**
 * Select n-levels deep into an object given a dot/bracket-notation query.
 * If partially applied, returns a function accepting the second argument.
 *
 * ### Examples:
 *
 *      selectn('name.first', contact);
 *
 *      selectn('addresses[0].street', contact);
 *
 *      contacts.map(selectn('name.first'));
 *
 * @param  {String | Array} query
 * dot/bracket-notation query string or array of properties
 *
 * @param  {Object} object
 * object to access
 *
 * @return {Function}
 * accessor function that accepts an object to be queried
 */

function selectn(query) {
  var parts;

  if (Array.isArray(query)) {
    parts = query;
  }
  else {
    // normalize query to `.property` access (i.e. `a.b[0]` becomes `a.b.0`)
    query = query.replace(/\[(\d+)\]/g, '.$1');
    parts = query.split('.');
  }

  /**
   * Accessor function that accepts an object to be queried
   *
   * @private
   *
   * @param  {Object} object
   * object to access
   *
   * @return {Mixed}
   * value at given reference or undefined if it does not exist
   */

  function accessor(object) {
    var ref = (object != null) ? object : (1, eval)('this');
    var len = parts.length;
    var idx = 0;

    // iteratively save each segment's reference
    for (; idx < len; idx += 1) {
      if (ref != null) ref = ref[parts[idx]];
    }

    return ref;
  }

  // curry accessor function allowing partial application
  return arguments.length > 1
       ? accessor(arguments[1])
       : accessor;
}

},{}]},{},[1])(1)
});
